### PR TITLE
feat: improve jsonschema validation on route resource

### LINKED ISSUE #2671: doc: for the `uri`, `host`, and `remote_add` fields of the route, plural field is used first
We need to add this tip in doc: https://github.com/apache/apisix/blob/master/doc/admin-api.md#route

Here is the code:

https://github.com/apache/apisix/blob/master/apisix/http/router/radixtree_uri.lua#L56

image

### LINKED ISSUE #2991: reject request data if it contains an empty array, eg: "remote_addrs": []
To add a route, if the `remote_addrs` field is an empty array, the matching behavior is ambiguous.

In order to avoid ambiguity and keep the API design simple and clear, we should reject the value as an empty array.

Here is an example, should return `400`:

```
$ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
{
    "uri": "/index.html",
    "remote_addrs": [],
    "enable_websocket": true,
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "39.97.63.215:80": 1
        }
    }
}'
HTTP/1.1 400 Bad Request
... ...
```

**Here is the todo list(from the discussion)**

* `Admin API`: reject request data if it contains an empty array, eg: `"remote_addrs": []`. `Code + Doc`
* `Manager API`: same as `Admin API`.
* `Dashbaord`: keep adding hacking codes to the frontend. `Code`

The field `hosts`, `uris` and `methods`, they have the same issue, we should fix them in the same way too.

original comment: https://github.com/apache/apisix/issues/2924#issuecomment-740405671

### PR DESCRIPTION
**Pull request by spacewander**
Fix #2671
Fix #2991

Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: doc/admin-api.md [hunk 1]
```diff
@@ -55,11 +55,12 @@
...
-|hosts    |False |Match Rules|The `host` in the form of a list means that multiple different hosts are allowed, and match any one of them.|{"foo.com", "*.bar.com"}|
-|remote_addr|False |Match Rules|The client requests an IP address: `192.168.1.101`, `192.168.1.102`, and CIDR format support `192.168.1.0/24`. In particular, APISIX also fully supports IPv6 address matching: `::1`, `fe80::1`, `fe80::1/64`, etc.|"192.168.1.0/24"|
-|remote_addrs|False |Match Rules|The `remote_addr` in the form of a list indicates that multiple different IP addresses are allowed, and match any one of them.|{"127.0.0.1", "192.0.0.0/8", "::1"}|
+|uri      |True, can't be used with `uris` |Match Rules|In addition to full matching such as `/foo/bar`、`/foo/gloo`, using different [Router](architecture-design.md#router) allows more advanced matching, see [Router](architecture-design.md#router) for more.|"/hello"|
+|uris     |True, can't be used with `uri`|Match Rules|The `uri` in the form of an empty list means that multiple different hosts are allowed, and match any one of them.|["/hello", "/word"]|
```
Comment (membphis) [inline doc/admin-api.md:59]: `different hosts are allowed` -> `different uris are allowed`

Comment (tokers) [reply doc/admin-api.md:59]: > The `uri` in the form of an empty list

Should be non-empty list.

File: doc/admin-api.md [hunk 2]
```diff
@@ -55,11 +55,12 @@
...
+|host     |False, can't be used with `hosts` |Match Rules|Currently requesting a domain name, such as `foo.com`; pan-domain names such as `*.foo.com` are also supported.|"foo.com"|
```
Comment (tokers) [inline doc/admin-api.md:60]: > pan-domain

Should be PAN domain.

File: doc/admin-api.md (hunk 1 above)
Comment (spacewander) [reply doc/admin-api.md:59]: Solved.

File: doc/admin-api.md (hunk 2 above)
Comment (spacewander) [reply doc/admin-api.md]: Solved.

**Review by tokers — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Comment by nic-chen** [comment]
@spacewander  missing update for `methods`..

**Comment by spacewander** [comment]
`methods` doesn't have its single form?

**Comment by nic-chen** [comment]
It doesn't have.
If passing an empty array for it does not affect matching route, we could ignore it.

**Pull request merged**